test/tools/config.py

```diff
@@ -0,0 +1,145 @@
+import yamale
+import random
+import string
+from os.path import dirname, abspath
+
+class Config:
+   def __init__(self):
+      self.dir = dirname(abspath(__file__))
+      self.base = dirname(dirname(dirname(abspath(__file__))))
+      self.schema = self.dir+'/schema.yaml'
+      self.params = {}
+      self.set_default_config()
+
+   def __getitem__(self, attr):
+      return self.params[attr]
+
+   def __setitem__(self, key, value):
+      self.params[key] = value
+
+   def set_default_config(self):
+      data = {
+            'org_id': None,
+            'project_id': None,
+            'project_name': self.generated_project_name(),
+            'device': self.default_device_config(),
+            'test': self.default_test_config()
+         }
+      self.params.update(data)
+
+   def set_common(self, org_id=None, key_name=None, dev_name=None):
+      if org_id is not None:
+         self.params['org_id'] = org_id
+      if key_name is not None:
+         self.params['device']['ssh_key_name'] = key_name
+      if dev_name is not None:
+         self.params['device']['name'] = dev_name
+
+   def set_run_flag_config(self, org_id=None, project_name=None, key_name=None, dev_name=None, skip_teardown=None):
+      self.set_common(org_id, key_name, dev_name)
+
+      if project_name is not None:
+         self.params['project_name'] = project_name
+      if skip_teardown is not None:
+         self.params['test']['skip_teardown'] = skip_teardown
+
+   def set_create_flag_config(self, org_id=None, project_id=None, key_name=None, dev_name=None):
+      self.set_common(org_id, key_name, dev_name)
+
+      if project_id is not None:
+         self.params['project_id'] = project_id
+
+   def load_config_from_file(self, config_file):
+      try:
+         schema = yamale.make_schema(self.schema)
+         data = yamale.make_data(config_file)
+         yamale.validate(schema, data)
+      except ValueError as e:
+         raise e
+      # This bit looks a bit more elaborate than it needs to be
+      # It seems that dict.update() does not work the way I need it to on
+      # nested dictionaries.
+      # For example, if device.userdata is set in dict1, but not in dict2,
+      # then the nested key will be removed completely.
+      # So I am updating each nested level separately. There is probably
+      # a smarter way to do this, but this is fine for now.
```

Comment:
why... does update (or the new merge thing which is in 3.9) not just do this? like if i want to update a dictionary i expect all of it to be updated? why is it an accepted thing for everyone to do workarounds?

Reply:
because a dict has no concept of the value type and the dict itself is not responsible to manage and check type. They are a simple key value store. If someone wants a more complex solution, it's usually better to implement it and reuse. Most of the configuration file handler libraries have their own Config object that handles this in a smart way.

Potentially we can have a `deep: bool` arg on update (with default False), but that would raise a lot of other questions like "can you implement a class that can be used with deep update/merge?" by design in python you would have this option, but than they have to extend all primitives to be other that class because than the signature would be not just Any but DeepReplaceable (or something better than this name). Does string, int, or boolean implements the new DeepReplaceable class? How can they?

It's kind of the same reason you have to use Reflect and a lot of hack in Go to deep compare/copy a struct that has a field with struct that has a field with struct. In go, you have to use reflect or use an existing library to do that. Python is the same, but the hack is much shorter because of the nature of the language.

Reply:
> It's kind of the same reason you have to use Reflect and a lot of hack in Go to deep compare/copy a struct that has a field with struct that has a field with struct. In go, you have to use reflect or use an existing library to do that. Python is the same, but the hack is much shorter because of the nature of the language.

yeh that makes sense, ty!